pyro/infer/mcmc/trace_kernel.py

```diff
@@ -0,0 +1,61 @@
+from __future__ import absolute_import, division, print_function
+
+from abc import ABCMeta, abstractmethod
+
+from six import add_metaclass
+
+
+@add_metaclass(ABCMeta)
+class TraceKernel(object):
+
+    def setup(self, *args, **kwargs):
+        """
+        Optional method to set up any state required before the run.
+
+        :param \*args: Algorithm specific positional arguments.
+        :param \*\*kwargs: Algorithm specific keyword arguments.
+        """
+        pass
+
+    def cleanup(self):
+        """
+        Optional method to clean up any residual state on termination.
+        """
+        pass
+
+    def diagnostics(self, time_step):
+        """
+        Relevant diagnostics (optional) to be printed at regular intervals
+        of the MCMC run. Returns `None` by default.
+
+        :param time_step: Current Monte Carlo time-step.
+        :return: String containing the diagnostic summary. e.g. acceptance rate
+        :rtype: string
+        """
+        return None
+
+    @abstractmethod
+    def initial_trace(self):
+        """
+        Returns an initial trace from the prior to initiate the MCMC run.
+
+        :return: Trace instance.
+        """
+        return NotImplementedError
+
+    @abstractmethod
+    def sample(self, trace):
```

Comment:
@eb8680, @fritzo - I simplified the method signature for `TraceKernel` and removed the args, kwargs arguments. Since we have a setup method, it should be able to take care of any other arguments that need to be provided to the TraceKernel instance to set up its state.